netbench-orchestrator/src/russula/mod.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use core::{task::Poll, time::Duration};
+use std::{collections::BTreeSet, net::SocketAddr};
+use tokio::net::TcpStream;
+use tracing::{error, info};
+
+mod error;
+mod event;
+mod network_utils;
+mod protocol;
+mod states;
+
+use error::{RussulaError, RussulaResult};
+use protocol::Protocol;
+
+const CONNENT_RETRY_ATTEMPT: usize = 10;
+
+#[derive(Debug, Copy, Clone)]
+pub enum ProtocolState {
+    /// The endpoint has established connection with its peer and
+    /// is ready to make progress.
+    Ready,
+
+    /// Indicates the protocol's terminal state.
+    Done,
+
+    /// Indicates that worker are running and accepting work.
+    ///
+    /// For netbench this state can be used to confirm that all servers are
+    /// running and accepting connection before starting netbench clients.
+    /// Should only be called by Coordinators.
+    WorkerRunning,
+}
+
+/// An established connection to a running instance of a Russula protocol.
+struct ProtocolInstance<P: Protocol> {
+    pub addr: SocketAddr,
+    pub stream: TcpStream,
+    pub protocol: P,
+}
+
+/// A Russula Endpoint.
+///
+/// An Endpoint can be of type Coordinator or Worker. A Coordinator can
+/// be used to synchronize multiple workers across different hosts. A Worker
+/// communicates with a Coordinator to make progress.
+pub struct Endpoint<P: Protocol> {
+    /// List of protocol instances to synchronize.
+    instance_list: Vec<ProtocolInstance<P>>,
+
+    /// Polling frequency.
+    poll_delay: Duration,
+}
+
+impl<P: Protocol + Send> Endpoint<P> {
+    /// Drive the underlying list of protocols until the desired state.
+    ///
+    /// For a non-blocking version use [Self::poll_until]
+    pub async fn run_untill(&mut self, desired_state: ProtocolState) -> RussulaResult<()> {
+        while self.poll_until(desired_state).await?.is_pending() {
+            tokio::time::sleep(self.poll_delay).await;
+        }
+
+        Ok(())
+    }
+
+    /// Poll the underlying list of protocols until the desired state.
+    ///
+    /// Attempt to drive each instance
+    pub async fn poll_until(&mut self, desired_state: ProtocolState) -> RussulaResult<Poll<()>> {
+        // Poll each peer protocol instance.
+        //
+        // If the peer is already in the desired state then this should be a noop.
+        for peer in self.instance_list.iter_mut() {
+            if let Err(err) = peer
+                .protocol
+                .poll_state(&mut peer.stream, desired_state)
+                .await
+            {
+                if err.is_fatal() {
+                    error!("{} {}", err, peer.addr);
+                    panic!("{} {}", err, peer.addr);
```

Comment:
Since the error is fatal, there is not much the application can do about it. An application would need to check if the error is fatal and shutdown/panic.

But its prob better to let the application handle it rather than panic.